packages/@aws-cdk/aws-redshift/lib/private/database-query-provider/table.ts

```diff
@@ -48,28 +71,79 @@ async function updateTable(
   }
 
   const oldTableColumns = oldResourceProperties.tableColumns;
   if (!oldTableColumns.every(oldColumn => tableColumns.some(column => column.name === oldColumn.name && column.dataType === oldColumn.dataType))) {
-    return createTable(tableNamePrefix, tableNameSuffix, tableColumns, clusterProps);
+    return createTable(tableNamePrefix, tableNameSuffix, tableColumns, tableAndClusterProps);
   }
 
-  const additions = tableColumns.filter(column => {
+  const columnAdditions = tableColumns.filter(column => {
     return !oldTableColumns.some(oldColumn => column.name === oldColumn.name && column.dataType === oldColumn.dataType);
   }).map(column => `ADD ${column.name} ${column.dataType}`);
-  await Promise.all(additions.map(addition => executeStatement(`ALTER TABLE ${tableName} ${addition}`, clusterProps)));
+  if (columnAdditions.length > 0) {
+    alterationStatements.push(...columnAdditions.map(addition => `ALTER TABLE ${tableName} ${addition}`));
+  }
+
+  const oldDistStyle = oldResourceProperties.distStyle;
+  if (!oldDistStyle && tableAndClusterProps.distStyle ||
+    oldDistStyle && !tableAndClusterProps.distStyle) {
+    return createTable(tableNamePrefix, tableNameSuffix, tableColumns, tableAndClusterProps);
+  } else if (oldDistStyle !== tableAndClusterProps.distStyle) {
+    alterationStatements.push(`ALTER TABLE ${tableName} ALTER DISTSTYLE ${tableAndClusterProps.distStyle}`);
+  }
+
+  const oldDistKey = getDistKeyColumn(oldTableColumns)?.name;
+  const newDistKey = getDistKeyColumn(tableColumns)?.name;
+  if (!oldDistKey && newDistKey ||
+    oldDistKey && !newDistKey
+  ) {
+    return createTable(tableNamePrefix, tableNameSuffix, tableColumns, tableAndClusterProps);
+  } else if (oldDistKey !== newDistKey) {
+    alterationStatements.push(`ALTER TABLE ${tableName} ALTER DISTKEY ${newDistKey}`);
+  }
+
+  const oldSortKeyColumns = getSortKeyColumns(oldTableColumns);
+  const newSortKeyColumns = getSortKeyColumns(tableColumns);
+  const oldSortStyle = oldResourceProperties.sortStyle;
+  const newSortStyle = tableAndClusterProps.sortStyle;
+  if ((oldSortStyle === newSortStyle && !areColumnsEqual(oldSortKeyColumns, newSortKeyColumns))
+    || (oldSortStyle !== newSortStyle)) {
+    switch (newSortStyle) {
+      case TableSortStyle.INTERLEAVED:
+        return createTable(tableNamePrefix, tableNameSuffix, tableColumns, tableAndClusterProps);
+
+      case TableSortStyle.COMPOUND: {
+        const sortKeyColumnsString = getSortKeyColumnsString(newSortKeyColumns);
+        alterationStatements.push(`ALTER TABLE ${tableName} ALTER ${newSortStyle} SORTKEY(${sortKeyColumnsString})`);
+        break;
+      }
+
+      case TableSortStyle.AUTO: {
+        alterationStatements.push(`ALTER TABLE ${tableName} ALTER SORTKEY ${newSortStyle}`);
+        break;
+      }
+    }
+  }
+
+  await Promise.all(alterationStatements.map(statement => executeStatement(statement, tableAndClusterProps)));
```

Comment:
I imagine there's a performance difference between executing a few alter table statements and a single one. According to the docs, sortkey/distkey changes can be combined into a single alter statement. Seems like it might be worth it to combine those here. What do you think?

Reply:
I agree that there would be some performance challenges but I think there would only be limited number of those statements and it is better to execute them separately to get appropriate error message if they fail.

Reply:
Fair enough.